RFC 128: Realms modifiers for Worlds (#128)
decentraland/adr: content/ADR-125-realm-modifiers-for-worlds-in-renderer.md

    version?: string
    commitHash?: string
    publicUrl: string
  }
  comms: {
    healthy: boolean
    protocol: string
    fixedAdapter?: string
    usersCount?: number
  }
  lambdas: {
    healthy: boolean
    version?: string
    commitHash?: string
    publicUrl: string
  }
  bff?: {
    healthy: boolean
    userCount: number
    commitHash?: string
    publicUrl: string
  }
}
```

Example:

```json
{
  "healthy": true,
  "content": {
    "healthy": true,
    "version": "5.1.2",
    "commitHash": "dce61c002c89db4966c0cdd008d3d654f297050b",
    "publicUrl": "https://peer.decentraland.org/content"
  },
  "configurations": {
    "realmName": "catalyst-4",
    "networkId": 1
  },
  "comms": {
    "healthy": true,
    "protocol": "v3",
    "fixedAdapter": "ws-room:mini-comms.decentraland.org/rooms/test-room"
  },
  "lambdas": {
    "healthy": true,
    "version": "5.1.2",
    "commitHash": "dce61c002c89db4966c0cdd008d3d654f297050b",
    "publicUrl": "https://peer.decentraland.org/lambdas"
  },
  "bff": {
    "healthy": true,
    "commitHash": "369c5dafeda62a1b16f5232cd477565cc3f3d513",
    "userCount": 0,
    "publicUrl": "/"
  }
}
```

## Current Implementation 

> <img alt="Realms Plugin UML" src="resources/ADR-125/Realms_Plugin_UML.png"/>

Reviewer: ```suggestion
```mermaid
classDiagram
   RealmPlugin <-- IRealmModifier
   IRealmModifier <|-- RealmBlockerModifier
   IRealmModifier <|-- RealmMinimapModifier
   class RealmPlugin {
      +List<IRealmModifier> realmModifiers
      +void SetRealmModifiers()
      +void RealmChanged(AboutResponse current, AboutResponse _)
    }
   class IRealmModifier {
        <<interface>>
   }
```
```

    usersCount?: number
  }
  lambdas: {
    healthy: boolean
    version?: string
    commitHash?: string
    publicUrl: string
  }
  bff?: {
    healthy: boolean
    userCount: number
    commitHash?: string
    publicUrl: string
  }
}
```

Example:

```json
{
  "healthy": true,
  "content": {
    "healthy": true,
    "version": "5.1.2",
    "commitHash": "dce61c002c89db4966c0cdd008d3d654f297050b",
    "publicUrl": "https://peer.decentraland.org/content"
  },
  "configurations": {
    "realmName": "catalyst-4",
    "networkId": 1
  },
  "comms": {
    "healthy": true,
    "protocol": "v3",
    "fixedAdapter": "ws-room:mini-comms.decentraland.org/rooms/test-room"
  },
  "lambdas": {
    "healthy": true,
    "version": "5.1.2",
    "commitHash": "dce61c002c89db4966c0cdd008d3d654f297050b",
    "publicUrl": "https://peer.decentraland.org/lambdas"
  },
  "bff": {
    "healthy": true,
    "commitHash": "369c5dafeda62a1b16f5232cd477565cc3f3d513",
    "userCount": 0,
    "publicUrl": "/"
  }
}
```

## Current Implementation 

![Untitled Diagram drawio](https://user-images.githubusercontent.com/1999557/202185605-65ebf318-a2bd-40c9-8908-200daa2166b5.png)

## Task breakdown 

> To complete the Beta version for worlds, the following modifiers are required:

Reviewer 2: Can you specify that it's part of Explorer task breakdown

Author: Done!

- **Skybox Modifier (V1 task):** Create a skybox modifier to allow the world creator control its state.
	- Skybox settings (time and dynamic toggle) will be unavaible while in world; they should be non-interactable.
	- A warning message should be added in settings to explicit to the user that the skybox is being controller by the creator
	- A specific time should be added to the AboutResponse.Configuration to allow the creators set the time of their World
	- Skybox should be set accordingly to the new variable mentioned above

> - **Green Blockers Modifiers:** Extra functionality should be added to the current green blockers modifier.

Reviewer 2: Should we use the same naming for consistency and replace `green blockers modifier` to  `Realms Blocker Modifier`

Author: Done!

	- Green blockers should appear 2 parcels away from the scene area

> - **Loading Screen Modifier (V1 task):** This are required functionalities that should be built on top of the current loading screen when accessing worlds. Requires coordination with the design team.

Reviewer 2: We might need to have a possibility for custom loading images.

Reviewer 2: That would require some design work, but just adding a background image would probably work.

Author: Added!

> - **Loading Screen Modifier (V1 task):** This are required functionalities that should be built on top of the current loading screen when accessing worlds. Requires coordination with the design team.

Reviewer 2: Do we need to disable the exiting tutorial from the flow for new users?

Author: Adding it as a task in the wizard section

```
namespace DCLPlugins.RealmsPlugin
{
    public interface IRealmsModifier : IDisposable
    {
        void OnEnteredRealm(bool isCatalyst, AboutResponse realmConfiguration);
    }
}
```

 The current modifiers currently implemented area:

- **Realms Blocker Modifier**

Controls how the green blockers should be shown on each realm. If a realm is a Catalyst, or it has a value on Configurations.CityLoaderContentServer; then the green blockers will be shown.

- **Realms Minimap Modifier**

Controls the functionality associated to the minimap when entering a Realm. It determines if the minimap should be visible using the realmConfiguration.Configurations.Minimap.Enabled value. Also, it determines if the Jump To Home Button should be present. This button will appear on any given user world, and it will take the user back to the home location in the most populated catalyst. 
When needed, this modifier will implement the change of the minimap image; as a user generated world will have the possibility to show its own minimap png.
	

### Data Structure 

> `AboutResponse` contains the info of the realm which has just been accessed. This class is autogenerated according to protobuff specifications. Each modifier gets access to the AboutResponse, and gets the info necessary to execute its behaviour. It has the following structure: 

Reviewer 3: It's not clear to me why renderer needs this `AboutResposne`.

Author: The `AboutResponse.configurations` contains the state necessary for the current modifiers. I will add an example to make it more clear; because I see that the one below doesn't have the `minimap` variable or `cityLoaderContentServer` variable; both are used in the current modifiers.

On the other hand, we discussed with @menduz sending the `AboutResponse` to replace the use of the `RealmsInfo`, since they are containing redundant info.  The redudant info that currently being sent is:

1) `content.publicURL` and `lambdas.publicURL`  are used to set the Catalyst info in renderer;
2) `configurations.realmName` is used for ExploreV2Menu to show the name of the current realm.

Right now they are co-existing, but whenever we want, we can remove the `UpdateRealmsInfo` call from Kernel.
![image](https://user-images.githubusercontent.com/1999557/203544916-bd4abc57-b7dc-4a69-b8c2-57b03672c497.png)

```typescript

> type About = {

Reviewer 4: This is duplicated info, maybe reference ADR-110 for clarity? https://adr.decentraland.org/adr/ADR-110

    version?: string
    commitHash?: string
    publicUrl: string
  }
  comms: {
    healthy: boolean
    protocol: string
    fixedAdapter?: string
    usersCount?: number
  }
  lambdas: {
    healthy: boolean
    version?: string
    commitHash?: string
    publicUrl: string
  }
  bff?: {
    healthy: boolean
    userCount: number
    commitHash?: string
    publicUrl: string
  }
}
```

Example:

```json
{
  "healthy": true,
  "content": {
    "healthy": true,
    "version": "5.1.2",
    "commitHash": "dce61c002c89db4966c0cdd008d3d654f297050b",
    "publicUrl": "https://peer.decentraland.org/content"
  },
  "configurations": {
    "realmName": "catalyst-4",
    "networkId": 1
  },
  "comms": {
    "healthy": true,
    "protocol": "v3",
    "fixedAdapter": "ws-room:mini-comms.decentraland.org/rooms/test-room"
  },
  "lambdas": {
    "healthy": true,
    "version": "5.1.2",
    "commitHash": "dce61c002c89db4966c0cdd008d3d654f297050b",
    "publicUrl": "https://peer.decentraland.org/lambdas"
  },
  "bff": {
    "healthy": true,
    "commitHash": "369c5dafeda62a1b16f5232cd477565cc3f3d513",
    "userCount": 0,
    "publicUrl": "/"
  }
}
```

## Current Implementation 

> ![Untitled Diagram drawio](https://user-images.githubusercontent.com/1999557/202185605-65ebf318-a2bd-40c9-8908-200daa2166b5.png)

Reviewer 4: the file needs to be included in the repo

	- Skybox settings (time and dynamic toggle) will be unavaible while in world; they should be non-interactable.
	- A warning message should be added in settings to explicit to the user that the skybox is being controller by the creator
	- A specific time should be added to the AboutResponse.Configuration to allow the creators set the time of their World
	- Skybox should be set accordingly to the new variable mentioned above

- **Realms Blocker Modifier:** Extra functionality should be added to the current green blockers modifier.
	- Green blockers should appear 2 parcels away from the scene area

- **Loading Screen Modifier (V1 task):** This are required functionalities that should be built on top of the current loading screen when accessing worlds. Requires coordination with the design team.
	- A message of entering world should be added
	- If a world is full (100 users), a warning message should appear (“The world you are trying to access is full. Please try again later”)
	- Add an image url for the loading screen to the AboutResponse json
	- If an image url exists, set it as background

- **Access World Modifier**

> 	- If user tries to go to an unexistent world, an empty world should be loaded

Reviewer: ```suggestion
	- If user tries to go to an inexistent world, an empty world should be loaded
```

> 	- If user tries to go to an unexistent world, an empty world should be loaded

Reviewer: Not sure how would this work... is this part of the about response? If it is, then the world does exist already. Otherwise the world content server should have 404'ed instead of returning a about response.

Reviewer: @pentreathm @menduz

Reviewer 5: Agreed, there is no such thing as an empty world unless it is an empty scene deployed to a NAME in the world server. Otherwise, if it doesn't exists, it should be managed by the client.

Reviewer 4: Can we have a configurable "fallback-world.dcl.eth" which can be deployed under the same rules as the rest? It will reduce one twist of configuration of the client and will enable common tools (cli, worlds server) to configure all the aspects of that empty parcel without requiring a release of servers or client itself. Wdyt?

Reviewer 5: I think it would be possible but as a user, if I type wrong the name of a world, I would expect the client to tell me that the world doesn't exist instead of taking me to an empty scene or we can make the fallback scene something fun and let users know that they are in a 404 World

Reviewer 4: Oh yeah, I love that approach, but that conflicts with the product definition. I definitely prefer the 404.

Reviewer 6: I love that approach too. @olavra ?

Reviewer 7: I love the idea of a 404 fallback scenes that we can build in the content team, and do crazy stuff. For the scope of the beta, I suggest to stick with the decentraland.org/404 page

	- A user should not be able to enter a world if it is full
	- Players should be able to enter a world using /goto world. (Right now we are accessing using changerealm command)

Besides these modifiers, we need to implement the following functionality. 

- **Teleport**
	- All teleports commands (goto random, crowd, coordinates, home) need to verify if we are on a DAO realm and if necessary change to one before completing the command. Its recommended to solve this on kernel directly.
	- The map will show Genesis City. As the teleport commands, we need to check realm before teleporting through the teleport menu and change if necessary.

- **Wizard**
	- As a new player enters DCL, it will be taken to this new tutorial wizard.
	- A world should never trigger the wizard.
	- New users are teleported back to Genesis City after the wizard
	- Remove the current tutorial for new users

> ## Task breakdown for SDK
> 
> This are the tasks that requires collaboration with the SDK team

Reviewer 4: When possible prevent talking about teams, projects are a better and more durable scope

```suggestion
## Changes for worlds-content-server

The [worlds-content-server](https://github.com/decentraland/worlds-content-server/) requires some changes to comply with this specification:
```
